File: tests/python/test_vision_ocr_preservation.py
"""
Preservation property tests for VisionRewardSystem OCR.

**Validates: Requirements 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7**

Property 2: Preservation — Non-Digit and Empty Input Behavior

These tests verify that extract_score() and compute_reward() behave
correctly for inputs where the bug condition does NOT hold: empty
observations, blank framebuffers, all-white framebuffers, random noise,
and reward delta / reset semantics.

These tests MUST PASS on the current UNFIXED code — they capture
baseline behavior that must be preserved after the fix.

Since VisionRewardSystem and extract_score() are not exposed to Python
bindings, we test through the compiled C++ GoogleTest binary which has
direct access to the class internals.
"""

import os
import subprocess
import sys

import pytest

# ---------------------------------------------------------------------------
# Path to the C++ test binary (built by cmake)
# ---------------------------------------------------------------------------
_REPO_ROOT = os.path.dirname(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
)
_BUILD_DIR = os.path.join(_REPO_ROOT, "build", "ci-linux")
_TEST_BINARY = os.path.join(_BUILD_DIR, "tests", "retro_ai_tests")


def _ensure_test_binary():
    """Build the C++ test binary if it doesn't exist."""
    if os.path.isfile(_TEST_BINARY):
        return
    result = subprocess.run(
        ["cmake", "--build", _BUILD_DIR, "--target", "retro_ai_tests", "-j4"],
        capture_output=True,
        text=True,
        timeout=120,
    )
    if result.returncode != 0:
        pytest.skip(f"Failed to build C++ test binary:\n{result.stderr}")


def _run_gtest(filter_pattern: str) -> subprocess.CompletedProcess:
    """Run the C++ test binary with a gtest filter and return the result."""
    _ensure_test_binary()
    return subprocess.run(
        [_TEST_BINARY, f"--gtest_filter={filter_pattern}"],
        capture_output=True,
        text=True,
        timeout=30,
    )


# ---------------------------------------------------------------------------
# Test 1: Empty observation → compute_reward() returns 0.0
#
# Validates: Requirement 3.1
# ---------------------------------------------------------------------------


def test_empty_observation_returns_zero_reward():
    """
    **Validates: Requirements 3.1**

    Empty observation → extract_score({}) returns -1 → compute_reward() returns 0.0.
    """
    result = _run_gtest("VisionOCRPreservation.EmptyObservationReturnsZeroReward")
    assert result.returncode == 0, (
        f"Empty observation preservation test FAILED.\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Test 2: Blank (all-zero) framebuffer → compute_reward() returns 0.0
#
# Validates: Requirement 3.2
# ---------------------------------------------------------------------------


def test_blank_framebuffer_returns_zero_reward():
    """
    **Validates: Requirements 3.2**

    All-zero (black) framebuffer → extract_score() returns -1 → reward = 0.0.
    """
    result = _run_gtest("VisionOCRPreservation.BlankFramebufferReturnsZeroReward")
    assert result.returncode == 0, (
        f"Blank framebuffer preservation test FAILED.\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Test 3: All-white framebuffer → compute_reward() returns 0.0
#
# Validates: Requirement 3.2
# ---------------------------------------------------------------------------


def test_all_white_framebuffer_returns_zero_reward():
    """
    **Validates: Requirements 3.2**

    All-255 (white) framebuffer → no contrast for Otsu → extract_score()
    returns -1 → reward = 0.0.
    """
    result = _run_gtest("VisionOCRPreservation.AllWhiteFramebufferReturnsZeroReward")
    assert result.returncode == 0, (
        f"All-white framebuffer preservation test FAILED.\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Test 4: Random noise framebuffers → compute_reward() returns 0.0
#
# Validates: Requirements 3.2, 3.3
# ---------------------------------------------------------------------------


@pytest.mark.parametrize(
    "seed",
    [42, 123, 999, 2024, 31415, 65535, 100000, 271828, 314159, 999999],
)
def test_random_noise_returns_zero_reward(seed: int):
    """
    **Validates: Requirements 3.2, 3.3**

    Random uint8 noise framebuffer → won't match ROM templates at 0.70
    threshold → extract_score() returns -1 → reward = 0.0.
    """
    result = _run_gtest(
        f"MultipleSeeds/VisionOCRPreservationNoise.RandomNoiseReturnsZeroReward/Seed{seed}"
    )
    assert result.returncode == 0, (
        f"Random noise preservation test FAILED (seed={seed}).\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Test 5: Reward delta preservation — no score on consecutive calls → 0.0
#
# Validates: Requirement 3.4
# ---------------------------------------------------------------------------


def test_reward_delta_preservation_no_score():
    """
    **Validates: Requirements 3.4**

    When no score is detected on consecutive calls, compute_reward()
    returns 0.0 each time.
    """
    result = _run_gtest("VisionOCRPreservation.RewardDeltaPreservationNoScore")
    assert result.returncode == 0, (
        f"Reward delta preservation test FAILED.\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Test 6: Reset preservation — after reset(), first compute_reward() returns 0.0
#
# Validates: Requirement 3.4
# ---------------------------------------------------------------------------


def test_reset_preservation():
    """
    **Validates: Requirements 3.4**

    After reset(), has_previous_ is false, so first compute_reward()
    returns 0.0 regardless of observation content.
    """
    result = _run_gtest("VisionOCRPreservation.ResetPreservation")
    assert result.returncode == 0, (
        f"Reset preservation test FAILED.\n"
        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
    )


# ---------------------------------------------------------------------------
# Aggregate: run ALL preservation tests at once
# ---------------------------------------------------------------------------


def test_all_preservation_tests():
    """
    **Validates: Requirements 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7**

    Run all C++ preservation tests. These MUST PASS on unfixed code.
    """
    result = _run_gtest("*Preservation*")
    assert result.returncode == 0, (
        f"Preservation tests FAILED — this is unexpected on unfixed code.\n"
        f"Test output summary:\n{result.stdout[-2000:]}"
    )

File: tests/test_vision_ocr_preservation.cpp
/**
 * @file test_vision_ocr_preservation.cpp
 * @brief Preservation property tests for VisionRewardSystem OCR.
 *
 * **Validates: Requirements 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7**
 *
 * Property 2: Preservation — Non-Digit and Empty Input Behavior
 *
 * These tests verify that extract_score() and compute_reward() behave
 * correctly for inputs where the bug condition does NOT hold: empty
 * observations, blank framebuffers, all-white framebuffers, random noise,
 * and reward delta / reset semantics.
 *
 * These tests MUST PASS on the current UNFIXED code — they capture
 * baseline behavior that must be preserved after the fix.
 */

#include <gtest/gtest.h>

#include "retro_ai/reward_systems/vision.hpp"
#include "retro_ai/rl_interface.hpp"

#include <cstdint>
#include <random>
#include <vector>

using namespace retro_ai;

// Videopac observation space: 160 height × 240 width × 3 channels (RGB888).
static constexpr int kObsHeight   = 160;
static constexpr int kObsWidth    = 240;
static constexpr int kObsChannels = 3;
static constexpr size_t kFramebufferSize =
    static_cast<size_t>(kObsHeight) * kObsWidth * kObsChannels;

// Default ScreenRegion used in the existing code.
static constexpr int kDefaultRegionX = 112;
static constexpr int kDefaultRegionY = 80;
static constexpr int kDefaultRegionW = 40;
static constexpr int kDefaultRegionH = 14;

// ---------------------------------------------------------------------------
// Helper: create a VisionRewardSystem with given score region.
// ---------------------------------------------------------------------------
static VisionRewardSystem make_vision_system(int rx, int ry, int rw, int rh) {
    ScreenRegion region{rx, ry, rw, rh};
    ObservationSpace obs{kObsWidth, kObsHeight, kObsChannels, 8};
    return VisionRewardSystem(region, obs);
}

// ---------------------------------------------------------------------------
// Helper: make a StepResult from a framebuffer.
// ---------------------------------------------------------------------------
static StepResult make_step(const std::vector<uint8_t>& fb) {
    StepResult s;
    s.observation = fb;
    s.reward = 0.0f;
    s.done = false;
    s.truncated = false;
    return s;
}

static StepResult make_empty_step() {
    return make_step({});
}

// ---------------------------------------------------------------------------
// Test 1: Empty observation → compute_reward() returns 0.0
//
// Validates: Requirement 3.1
// extract_score({}) returns -1 → compute_reward returns 0.0
// ---------------------------------------------------------------------------
TEST(VisionOCRPreservation, EmptyObservationReturnsZeroReward) {
    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);
    vision.reset();

    auto step = make_empty_step();
    StepResult dummy;
    dummy.observation = {};

    float reward = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(reward, 0.0f)
        << "Empty observation should produce 0.0 reward (extract_score returns -1).";
}

// ---------------------------------------------------------------------------
// Test 2: Blank (all-zero / black) framebuffer → compute_reward() returns 0.0
//
// Validates: Requirement 3.2
// All-zero pixels → extract_score returns -1 → reward = 0.0
// ---------------------------------------------------------------------------
TEST(VisionOCRPreservation, BlankFramebufferReturnsZeroReward) {
    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);
    vision.reset();

    std::vector<uint8_t> fb(kFramebufferSize, 0);
    auto step = make_step(fb);
    StepResult dummy;
    dummy.observation = {};

    float reward = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(reward, 0.0f)
        << "Blank (all-black) framebuffer should produce 0.0 reward.";
}

// ---------------------------------------------------------------------------
// Test 3: All-white framebuffer → compute_reward() returns 0.0
//
// Validates: Requirement 3.2
// All-255 pixels → Otsu threshold has no contrast → extract_score returns -1
// ---------------------------------------------------------------------------
TEST(VisionOCRPreservation, AllWhiteFramebufferReturnsZeroReward) {
    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);
    vision.reset();

    std::vector<uint8_t> fb(kFramebufferSize, 255);
    auto step = make_step(fb);
    StepResult dummy;
    dummy.observation = {};

    float reward = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(reward, 0.0f)
        << "All-white framebuffer should produce 0.0 reward (no contrast for Otsu).";
}


// ---------------------------------------------------------------------------
// Test 4: Random noise framebuffer → compute_reward() returns 0.0
//
// Validates: Requirements 3.2, 3.3
// Random uint8 noise won't match ROM templates at 0.70 threshold.
// We test multiple seeds to increase confidence.
// ---------------------------------------------------------------------------
class VisionOCRPreservationNoise : public ::testing::TestWithParam<uint32_t> {};

TEST_P(VisionOCRPreservationNoise, RandomNoiseReturnsZeroReward) {
    uint32_t seed = GetParam();

    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);
    vision.reset();

    std::mt19937 rng(seed);
    std::uniform_int_distribution<int> dist(0, 255);

    std::vector<uint8_t> fb(kFramebufferSize);
    for (auto& pixel : fb) {
        pixel = static_cast<uint8_t>(dist(rng));
    }

    auto step = make_step(fb);
    StepResult dummy;
    dummy.observation = {};

    float reward = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(reward, 0.0f)
        << "Random noise framebuffer (seed=" << seed
        << ") should produce 0.0 reward (no digit patterns at 0.70 threshold).";
}

INSTANTIATE_TEST_SUITE_P(
    MultipleSeeds,
    VisionOCRPreservationNoise,
    ::testing::Values(42, 123, 999, 2024, 31415, 65535, 100000, 271828, 314159, 999999),
    [](const ::testing::TestParamInfo<uint32_t>& info) {
        return "Seed" + std::to_string(info.param);
    });

// ---------------------------------------------------------------------------
// Test 5: Reward delta preservation — no score on consecutive calls → 0.0
//
// Validates: Requirement 3.4
// When no score is detected on consecutive calls, compute_reward() returns 0.0.
// ---------------------------------------------------------------------------
TEST(VisionOCRPreservation, RewardDeltaPreservationNoScore) {
    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);
    vision.reset();

    std::vector<uint8_t> fb(kFramebufferSize, 0);
    auto step = make_step(fb);
    StepResult dummy;
    dummy.observation = {};

    // Multiple consecutive calls with blank framebuffer
    float r1 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r1, 0.0f) << "First call with blank fb should return 0.0";

    float r2 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r2, 0.0f) << "Second call with blank fb should return 0.0";

    float r3 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r3, 0.0f) << "Third call with blank fb should return 0.0";
}

// ---------------------------------------------------------------------------
// Test 6: Reset preservation — after reset(), first compute_reward() returns 0.0
//
// Validates: Requirement 3.4
// After reset(), has_previous_ is false, so first detection returns 0.0.
// ---------------------------------------------------------------------------
TEST(VisionOCRPreservation, ResetPreservation) {
    auto vision = make_vision_system(
        kDefaultRegionX, kDefaultRegionY, kDefaultRegionW, kDefaultRegionH);

    // First cycle
    vision.reset();

    std::vector<uint8_t> fb(kFramebufferSize, 0);
    auto step = make_step(fb);
    StepResult dummy;
    dummy.observation = {};

    float r1 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r1, 0.0f) << "First call after reset should return 0.0";

    // Reset again
    vision.reset();

    float r2 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r2, 0.0f) << "First call after second reset should return 0.0";

    // Third reset
    vision.reset();

    float r3 = vision.compute_reward(step, dummy);
    EXPECT_FLOAT_EQ(r3, 0.0f) << "First call after third reset should return 0.0";
}
